Add export to WAV command-line functionality

trh-x opened the pull request:
This PR introduces the ability to export to WAV from the command line. It adds the capability to MilkyTracker and also adds an independent CLI utility called `export-to-wav`.

It also adds a general-purpose CLI parser which can be used to add more options/commands in the future.

## Multi-track

The CLI functionality includes the ability to export in multi-track mode. When doing so, WAVs which are silent will automatically be removed.

# Usage

## MilkyTracker (MacOS Cocoa)

```
Usage: ./build/src/tracker/MilkyTracker.app/Contents/MacOS/MilkyTracker [input] [options]

Positional arguments:
  input                Input module file (.xm) (optional)

Options:
  -help               Show this help message and exit
  -headless           Run in headless mode
  -output <value>     Output file name
  -sample-rate <value>Sample rate in Hz (default: from settings or 44100)
  -volume <value>     Mixer volume (default: from settings or 256)
  -shift <value>      Mixer shift (default: from settings or 1)
  -resampler <value>  Resampler type (default: from settings or 4)
  -multi-track        Export each track to a separate WAV file
  -verbose            Enable verbose output

When using -multi-track, output files will be named:
  output_01.wav, output_02.wav, etc.
  (Silent tracks will be automatically removed)
```

## MilkyTracker (SDL)

```
Usage: ./build/src/tracker/milkytracker [input] [options]

Positional arguments:
  input                Input module file (.xm) (optional)

Options:
  -help               Show this help message and exit
  -bpp <value>        Set bits per pixel
  -nosplash           Skip splash screen
  -swap               Swap red and blue colors
  -orientation <value>Set screen orientation (allowed values: NORMAL, ROTATE90CCW, ROTATE90CW)
  -recvelocity        Enable recording velocity
  -headless           Run in headless mode
  -output <value>     Output file name
  -sample-rate <value>Sample rate in Hz (default: from settings or 44100)
  -volume <value>     Mixer volume (default: from settings or 256)
  -shift <value>      Mixer shift (default: from settings or 1)
  -resampler <value>  Resampler type (default: from settings or 4)
  -multi-track        Export each track to a separate WAV file
  -verbose            Enable verbose output

When using -multi-track, output files will be named:
  output_01.wav, output_02.wav, etc.
  (Silent tracks will be automatically removed)
```

## `export-to-wav`

```
Usage: ./build/src/tools/export-to-wav/export-to-wav <input> [options]

Positional arguments:
  input                Input module file (.xm)

Options:
  -help               Show this help message and exit
  -output <value>     Output file name
  -sample-rate <value>Sample rate in Hz (default: from settings or 44100)
  -volume <value>     Mixer volume (default: from settings or 256)
  -shift <value>      Mixer shift (default: from settings or 1)
  -resampler <value>  Resampler type (default: from settings or 4)
  -multi-track        Export each track to a separate WAV file
  -verbose            Enable verbose output

When using -multi-track, output files will be named:
  output_01.wav, output_02.wav, etc.
  (Silent tracks will be automatically removed)
Error: Missing required argument: input
```

# Notes

A few other changes are included, some of which might be preferable to omit or merge separately:

- Add `commandlib` library to allow reusing CLI code in the tracker and the `export-to-wav` utility.
- Switch C++ standard to C++11 (see https://github.com/milkytracker/MilkyTracker/pull/328).
- Add functionality to `build.sh`:
  - Allow a debug build to be built.
  - Allow DMG generation to be skipped on Mac.

trh-x commented:
Hey @coderofsalvation! Thanks for your interest and feedback on this.

> Very cool! Btw. have you seen: [31ba8a9](https://github.com/milkytracker/MilkyTracker/commit/31ba8a91fbaafdc285cf3ab4792c978cd287d3a9)

It's great to see the multi channel export land in the tracker. I must confess, I am not actively using MilkyTracker at the moment, at least not in a compositional sense - my position is that I've got a load of old XM files I wrote on FT2 in the mid/late 90's and I want to automate exporting them to WAVs to make them more accessible. Hence the initial `export-to-wav` proposal. But I've been very glad to find the spiritual successor of FT2 is going strong and in active development!

> I really like love the idea of having this available in a scriptable/batch way (for example to use in combination with [faircamp](https://simonrepp.com/faircamp/))

Yup, as you've gathered that's my use case - thanks for the heads up on Faircamp, I've not come across that before. Looks very cool.

I'm in the process of putting together a site myself which allows uploading and auto-exporting mods to their multi-track components. It's called [wavtopia](https://wavtopia.com), very much a work in progress at the moment. As you can [see](https://github.com/trh-x/wavtopia/blob/f086aa309fa3bbd47d4da776260ad6845d787760/packages/media/src/config.ts#L10), I'm already making use of the CLI utility.

> questions:
> 
> 1. Could this be also added to the `milkytracker` binary (right now it just detects one arguments, and loads it as a module).
> 2. should we rename `trackerlib` to `commandlib` or `clilib` or something to make it more specific
> 
> > Reason: I've been thinking of a cli interface for milkytracker before, and I can think of more usecases (besides exporting). For example: we have some env-vars (`MIDI_IN` and `SCALE_NEAREST`) but they are very hidden in the docs.
> 
> So WDYT about this:
> 
> ...
> 
> Then later --midi-in and --scale-nearest can be added as args, and who knows what.

That all sounds great and I'm glad to see you'd like to bring this into the tracker itself. I was considering that at the start but I figured it would be a safer bet to keep things as light touch as possible.

But, I've now gone ahead and integrated CLI parsing and WAV exporting into the tracker. I noticed the SDL version already had some CLI arguments, I've replaced their parsing with the CLI parser that I've put together.

The SDL argument format uses a single dash prefix but I've kept the double dash for the utility and the Cocoa version, but maybe it would make more sense to update these all to use the same format. WDYT?

trh-x commented:
Hi @coderofsalvation, just thinking more on this, maybe we don't even need the separate `export-to-wav` utility, and could instead just use MilkyTracker with the `--headless` flag.

My thinking in retaining the separate executable was that it could potentially allow `export-to-wav` to be installed on a system with less dependencies than the full tracker.

However, taking a look at ldd output to get a measure of this:

<details>
<summary>
ldd build/src/tracker/milkytracker
</summary>

```
	/lib64/ld-linux-x86-64.so.2 (0x000071f761e4b000)
	libapparmor.so.1 => /lib/x86_64-linux-gnu/libapparmor.so.1 (0x000071f760b7b000)
	libasound.so.2 => /lib/x86_64-linux-gnu/libasound.so.2 (0x000071f7619d8000)
	libasyncns.so.0 => /lib/x86_64-linux-gnu/libasyncns.so.0 (0x000071f760b8f000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x000071f760676000)
	libcap.so.2 => /lib/x86_64-linux-gnu/libcap.so.2 (0x000071f7608e4000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000071f761253000)
	libdbus-1.so.3 => /lib/x86_64-linux-gnu/libdbus-1.so.3 (0x000071f760d9b000)
	libdecor-0.so.0 => /lib/x86_64-linux-gnu/libdecor-0.so.0 (0x000071f760e68000)
	libdrm.so.2 => /lib/x86_64-linux-gnu/libdrm.so.2 (0x000071f760eee000)
	libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x000071f760d23000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x000071f760d06000)
	libFLAC.so.12 => /lib/x86_64-linux-gnu/libFLAC.so.12 (0x000071f760b07000)
	libgbm.so.1 => /lib/x86_64-linux-gnu/libgbm.so.1 (0x000071f760edd000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x000071f761465000)
	libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x000071f76079c000)
	libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x000071f760651000)
	liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1 (0x000071f760778000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x000071f760746000)
	libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x000071f760640000)
	libmp3lame.so.0 => /lib/x86_64-linux-gnu/libmp3lame.so.0 (0x000071f7608f1000)
	libmpg123.so.0 => /lib/x86_64-linux-gnu/libmpg123.so.0 (0x000071f760967000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000071f761495000)
	libogg.so.0 => /lib/x86_64-linux-gnu/libogg.so.0 (0x000071f7609c3000)
	libopus.so.0 => /lib/x86_64-linux-gnu/libopus.so.0 (0x000071f7609cf000)
	libpulsecommon-16.1.so => /usr/lib/x86_64-linux-gnu/pulseaudio/libpulsecommon-16.1.so (0x000071f760dea000)
	libpulse.so.0 => /lib/x86_64-linux-gnu/libpulse.so.0 (0x000071f761202000)
	libsamplerate.so.0 => /lib/x86_64-linux-gnu/libsamplerate.so.0 (0x000071f761093000)
	libSDL2-2.0.so.0 => /lib/x86_64-linux-gnu/libSDL2-2.0.so.0 (0x000071f7617fc000)
	libsndfile.so.1 => /lib/x86_64-linux-gnu/libsndfile.so.1 (0x000071f760c7e000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x000071f76157e000)
	libsystemd.so.0 => /lib/x86_64-linux-gnu/libsystemd.so.0 (0x000071f760b97000)
	libvorbisenc.so.2 => /lib/x86_64-linux-gnu/libvorbisenc.so.2 (0x000071f760a2e000)
	libvorbis.so.0 => /lib/x86_64-linux-gnu/libvorbis.so.0 (0x000071f760ad9000)
	libwayland-client.so.0 => /lib/x86_64-linux-gnu/libwayland-client.so.0 (0x000071f760ec8000)
	libwayland-cursor.so.0 => /lib/x86_64-linux-gnu/libwayland-cursor.so.0 (0x000071f760ebe000)
	libwayland-egl.so.1 => /lib/x86_64-linux-gnu/libwayland-egl.so.1 (0x000071f760ed8000)
	libwayland-server.so.0 => /lib/x86_64-linux-gnu/libwayland-server.so.0 (0x000071f760d4e000)
	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x000071f760f56000)
	libX11-xcb.so.1 => /lib/x86_64-linux-gnu/libX11-xcb.so.1 (0x000071f760c77000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x000071f760b75000)
	libxcb-randr.so.0 => /lib/x86_64-linux-gnu/libxcb-randr.so.0 (0x000071f760d12000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x000071f760d72000)
	libXcursor.so.1 => /lib/x86_64-linux-gnu/libXcursor.so.1 (0x000071f760f34000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x000071f760b6b000)
	libXext.so.6 => /lib/x86_64-linux-gnu/libXext.so.6 (0x000071f760f40000)
	libXfixes.so.3 => /lib/x86_64-linux-gnu/libXfixes.so.3 (0x000071f760f19000)
	libXi.so.6 => /lib/x86_64-linux-gnu/libXi.so.6 (0x000071f760f21000)
	libxkbcommon.so.0 => /lib/x86_64-linux-gnu/libxkbcommon.so.0 (0x000071f760e75000)
	libXrandr.so.2 => /lib/x86_64-linux-gnu/libXrandr.so.2 (0x000071f760f0c000)
	libXrender.so.1 => /lib/x86_64-linux-gnu/libXrender.so.1 (0x000071f760d66000)
	libXss.so.1 => /lib/x86_64-linux-gnu/libXss.so.1 (0x000071f760f07000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x000071f761ae2000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x000071f76068c000)
	linux-vdso.so.1 (0x00007fff5f34e000)
```
</details>

<details>
<summary>
ldd build/src/tools/export-to-wav/export-to-wav
</summary>

```
	linux-vdso.so.1 (0x00007ffdbe16d000)
	libasound.so.2 => /lib/x86_64-linux-gnu/libasound.so.2 (0x00007c16d3caf000)
	libSDL2-2.0.so.0 => /lib/x86_64-linux-gnu/libSDL2-2.0.so.0 (0x00007c16d3ad3000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007c16d3855000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007c16d376c000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007c16d373e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007c16d352a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007c16d3e55000)
	libpulse.so.0 => /lib/x86_64-linux-gnu/libpulse.so.0 (0x00007c16d34d9000)
	libsamplerate.so.0 => /lib/x86_64-linux-gnu/libsamplerate.so.0 (0x00007c16d336a000)
	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007c16d322d000)
	libXext.so.6 => /lib/x86_64-linux-gnu/libXext.so.6 (0x00007c16d3219000)
	libXcursor.so.1 => /lib/x86_64-linux-gnu/libXcursor.so.1 (0x00007c16d320d000)
	libXi.so.6 => /lib/x86_64-linux-gnu/libXi.so.6 (0x00007c16d31f8000)
	libXfixes.so.3 => /lib/x86_64-linux-gnu/libXfixes.so.3 (0x00007c16d31f0000)
	libXrandr.so.2 => /lib/x86_64-linux-gnu/libXrandr.so.2 (0x00007c16d31e3000)
	libXss.so.1 => /lib/x86_64-linux-gnu/libXss.so.1 (0x00007c16d31de000)
	libdrm.so.2 => /lib/x86_64-linux-gnu/libdrm.so.2 (0x00007c16d31c7000)
	libgbm.so.1 => /lib/x86_64-linux-gnu/libgbm.so.1 (0x00007c16d31b6000)
	libwayland-egl.so.1 => /lib/x86_64-linux-gnu/libwayland-egl.so.1 (0x00007c16d31af000)
	libwayland-client.so.0 => /lib/x86_64-linux-gnu/libwayland-client.so.0 (0x00007c16d319f000)
	libwayland-cursor.so.0 => /lib/x86_64-linux-gnu/libwayland-cursor.so.0 (0x00007c16d3195000)
	libxkbcommon.so.0 => /lib/x86_64-linux-gnu/libxkbcommon.so.0 (0x00007c16d314c000)
	libdecor-0.so.0 => /lib/x86_64-linux-gnu/libdecor-0.so.0 (0x00007c16d3141000)
	libpulsecommon-16.1.so => /usr/lib/x86_64-linux-gnu/pulseaudio/libpulsecommon-16.1.so (0x00007c16d30c1000)
	libdbus-1.so.3 => /lib/x86_64-linux-gnu/libdbus-1.so.3 (0x00007c16d3072000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007c16d3049000)
	libXrender.so.1 => /lib/x86_64-linux-gnu/libXrender.so.1 (0x00007c16d303d000)
	libwayland-server.so.0 => /lib/x86_64-linux-gnu/libwayland-server.so.0 (0x00007c16d3027000)
	libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007c16d2ffa000)
	libxcb-randr.so.0 => /lib/x86_64-linux-gnu/libxcb-randr.so.0 (0x00007c16d2fe9000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007c16d2fdd000)
	libsndfile.so.1 => /lib/x86_64-linux-gnu/libsndfile.so.1 (0x00007c16d2f55000)
	libX11-xcb.so.1 => /lib/x86_64-linux-gnu/libX11-xcb.so.1 (0x00007c16d2f50000)
	libsystemd.so.0 => /lib/x86_64-linux-gnu/libsystemd.so.0 (0x00007c16d2e6e000)
	libasyncns.so.0 => /lib/x86_64-linux-gnu/libasyncns.so.0 (0x00007c16d2e66000)
	libapparmor.so.1 => /lib/x86_64-linux-gnu/libapparmor.so.1 (0x00007c16d2e52000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007c16d2e4c000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007c16d2e44000)
	libFLAC.so.12 => /lib/x86_64-linux-gnu/libFLAC.so.12 (0x00007c16d2dde000)
	libvorbis.so.0 => /lib/x86_64-linux-gnu/libvorbis.so.0 (0x00007c16d2db0000)
	libvorbisenc.so.2 => /lib/x86_64-linux-gnu/libvorbisenc.so.2 (0x00007c16d2d05000)
	libopus.so.0 => /lib/x86_64-linux-gnu/libopus.so.0 (0x00007c16d2ca6000)
	libogg.so.0 => /lib/x86_64-linux-gnu/libogg.so.0 (0x00007c16d2c9c000)
	libmpg123.so.0 => /lib/x86_64-linux-gnu/libmpg123.so.0 (0x00007c16d2c40000)
	libmp3lame.so.0 => /lib/x86_64-linux-gnu/libmp3lame.so.0 (0x00007c16d2bc8000)
	libcap.so.2 => /lib/x86_64-linux-gnu/libcap.so.2 (0x00007c16d2bbb000)
	libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x00007c16d2a73000)
	liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1 (0x00007c16d2a51000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007c16d2a1f000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007c16d2963000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007c16d294d000)
	libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x00007c16d2928000)
	libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007c16d2919000)
```
</summary>
</details>

At present, the only difference is that `export-to-wav` doesn't require `libz`.

Maybe this is something that could be improved, and so worth keeping `export-to-wav` for, or maybe it's not really worth it. What do you reckon?

coderofsalvation commented:
>  old XM files I wrote on FT2 in the late 90's and I want to automate exporting them to WAVs to make them more accessible. Hence the initial export-to-wav proposal. 

Sounds familiar. There's a vast universe out there of unreleased MODS/XMS etc, with lots of listening and remix potential.
Recently I've remixed an XM of my buddy from 96: https://www.youtube.com/watch?v=VJ5ONO2oqlk

> But I've been very glad to find the spiritual successor of FT2 is going strong and in active development.

Thx! With all the modern DAW's at our disposal, a longing for less plugins & simplicity has resparked a wide interest in trackers again. Luckily modules don't break over time (unlike fileformats with plugin-dependencies).

> thanks for the heads up on Faircamp as well, I've not come across that before. Looks very cool.

NP, I've talked to the developer quite a bit, great guy.

> t's called [wavtopia](https://wavtopia.com/), very much a work in progress at the moment. As you can [see](https://github.com/trh-x/wavtopia/blob/f086aa309fa3bbd47d4da776260ad6845d787760/packages/media/src/config.ts#L10), I'm already making use of the CLI utility :)

I've added a feature-request to the wavtopia repo (superexcited to see this new effort!)

> The SDL argument format uses a single dash prefix but I've kept the double dash for the utility and the Cocoa version, but maybe it would make more sense to update these all to use the same format. WDYT?

Makes sense, I really love when CLI cmds are copy/paste-able across platforms without single/double-dash rewrites.

> Update: Actually, something else has occurred to me, which is that it might be a good idea to bring the behaviour whereby silent tracks are deleted into the GUI export flow to make it the standard behaviour. But that's probably better tackled in a separate PR

Good point, agreed.

> Hi @coderofsalvation, just thinking more on this, maybe we don't even need the separate export-to-wav utility, and could instead just use MilkyTracker with the --headless flag.

I really like the idea, do you mean that the `--headless` flag could skip Window/Video-stuff initialisations, so it can also work on a server (which does not have all the appropriate video/audio dependencies) ?

trh-x commented:
Hey @coderofsalvation! Hope you're keeping well.

> Sounds familiar. There's a vast universe out there of unreleased MODS/XMS etc, with lots of listening and remix potential. Recently I've remixed an XM of my buddy from 96: [youtube.com/watch?v=VJ5ONO2oqlk](https://www.youtube.com/watch?v=VJ5ONO2oqlk)

Just had a listen to that - nice track! And it's great to see a module from '96 getting a remix in the current era. It really showcases the stability of the mod format.

> > But I've been very glad to find the spiritual successor of FT2 is going strong and in active development.
> 
> Thx! With all the modern DAW's at our disposal, a longing for less plugins & simplicity has resparked a wide interest in trackers again. Luckily modules don't break over time (unlike fileformats with plugin-dependencies).

Yup it's easy to see the appeal of a simpler approach, compared to the endless number of plugins and complexity of the modern DAWs. Nice to know trackers are on the up!

> I've added a feature-request to the wavtopia repo (superexcited to see this new effort!). I'd love to see more click-to-play URLs which link to wavtopia. Its worth exploring multichannel music, especially with established formats. PIONEER on the other hand, unsuccesfully attempted to promote multichannel music ('stems') not long ago. Needless to say, its adoption among composers/listeners never came near the MOD/XM/S3M/IT-scene (perhaps the slow modem-ERA and pre-MP3-ERA were essential ingredients for it to take off).

That's awesome, thanks for the great suggestion for wavtopia! Definitely one I'll get to in time (or anyone else who feels like getting involved of course).

Interesting to her about Pioneer's efforts, I didn't know about that. I'll have to look it up, just to see what sort of angle they were taking with it even if it didn't take off...

> > The SDL argument format uses a single dash prefix but I've kept the double dash for the utility and the Cocoa version, but maybe it would make more sense to update these all to use the same format. WDYT?
> 
> Makes sense, I really love when CLI cmds are copy/paste-able across platforms without single/double-dash rewrites.

I've updated the CLI args to use single dashes for all cases, for consistency. Actually, I must admit I find it a little odd, and do prefer a double dash for full arg names vs a single dash for abbreviated names (`--help`, `-h` etc). But, with the SDL version already present with single dash args, it seems to makes sense to stick with that, rather than introducing a breaking change.

> > Update: Actually, something else has occurred to me, which is that it might be a good idea to bring the behaviour whereby silent tracks are deleted into the GUI export flow to make it the standard behaviour. But that's probably better tackled in a separate PR
> 
> Good point, agreed. Another thing I'd like to adress at some point is un-spaghettified-export of modules (because many mods have multiple instruments on 1 track). It would be nice to export one WAV-file per instrument, so it's easier to mix/master in another environment.

That's an interesting point. I think some tracks would benefit from that, while others might have multiple instruments on a track but they are best grouped within that track. Would definitely be a good option to have!

> > Hi @coderofsalvation, just thinking more on this, maybe we don't even need the separate export-to-wav utility, and could instead just use MilkyTracker with the --headless flag.
> 
> I really like the idea, do you mean that the `--headless` flag could skip Window/Video-stuff initialisations, so it can also work on a server (which does not have all the appropriate video/audio dependencies) ?

So, yeah the `-headless` flag is already there and will skip the GUI initialisation, so it should be server friendly.

That said - on reflection, I think there's still a case to be made for keeping the separate `export-to-wav` executable.

Comparing the file size of `export-to-wav` vs the tracker, you can see that `export-to-wav` is quite a lot smaller.  Although maybe it's not such a big deal in this day and age, it's still nice to have the option for a slimmed down executable that's more server friendly.

Also, as I mentioned on my previous comment while there's currently not much difference in the shared library dependencies between the two executables, this is still something that can be improved such that `export-to-wav` is deployable on a system with a lot less requirements compared to the full tracker.

Lastly, thinking about removing `export-to-wav` had me thinking about reverting the `commandlib` CMake refactor; this seems like something it would be nice to develop and improve rather than remove. Interested to hear your thoughts!

```
# Linux
$ ls -l build/src/tracker/milkytracker build/src/tools/export-to-wav/export-to-wav | awk '{print $5, $9}'
610008 build/src/tools/export-to-wav/export-to-wav
3544608 build/src/tracker/milkytracker

# Mac
$ ls -l build/src/tracker/MilkyTracker.app/Contents/MacOS/MilkyTracker build/src/tools/export-to-wav/export-to-wav | awk '{print $5, $9}'
501792 build/src/tools/export-to-wav/export-to-wav
2979120 build/src/tracker/MilkyTracker.app/Contents/MacOS/MilkyTracker
```

coderofsalvation commented:
All good points.
Btw. quick observation: what do you think of renaming `export-to-wav` to `milkycli`? 
It would be a bit more descriptive, especially on servers.
It would also allow for adding future cli subcmds, without having multiple executables which are 99% the same.

trh-x commented:
> All good points. Btw. quick observation: what do you think of renaming `export-to-wav` to `milkycli`? It would be a bit more descriptive, especially on servers. It would also allow for adding future cli subcmds, without having multiple executables which are 99% the same.

That sounds sensible, I've renamed it to `milkycli` as per your suggestion.
